### DISCUSSION TITLE
Print statement delays execution.

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#9502] [General]
I have a simple micropython program that begins execution when an external device enables the ESP8266 chip. Because the external signal only lasts for 2 seconds, one of the first things the code does is output a signal to maintain the "EN" pin during code execution. There are several debugging print statements that follow the ENable-hold signal. When the code completes it's task, it signifies by flashing a LED. In standalone (no IDE), it sporadically responds to the external signal; sometimes flashing the LED, sometimes not, indicating that for some reason the ENable-hold wasn't holding. I noticed when used in the IDE there's was a delay before any execution of 500-1000ms, I suspected the print statements and commented them out. With the print statements commented out, code execution was immediate, and in standalone the LED flashed with every external signal; the ENable-hold was working every time.

Given that the ENable-hold was executed before any print statement, does anyone have any thoughts as to why it's affected by the print statement, and why the print statements appear to delay code execution?

**Comment by robert-hh** [comment]
Print through UBS/Serial is pretty straight forward at the ESP8266. Once a print started, it will block until all characters are sent. So it might be that the print which block an event is resulting from a previous event, depending on how fast the events arrive. But not knowing your code that's just guessing. I do not know of you process the input event by a Pin IRQ or by polling. In any case, the time to print the debug messages must be shorter that the time between events.
One question: The "EN" signal you're talking about, is that the ESP8266 or one of the external device?
A note: IDE's can make things worse, since they add own communication. If it is about timing, better use simple terminal emulators and/or logic analyzers. Even the simple 12$ devices are of exceptional  help.

**Comment by mmliam** [comment]
Hi Robert,
Thanks for the reply. My thought was exactly your supposition; the print statement is suspending, or interrupting the ENable-hold (ESP8266 GPIO4). In standalone I noticed that it is raised, but gets interrupted when the print statements are not commented out. The "EN" is the pin on the ESP8266.

**Reply by robert-hh** [reply]
Why do you touch "EN". Usually, this is kept open, unless you want to switch the ESP8266 on and off.

**Reply by mmliam** [reply]
For my application, I need to turn the ESP8266 on and off.

**Comment by davefes** [comment]
> one of the first things the code does is output a signal to maintain the "EN" pin during code execution.

Which pin do you output a signal on?  Search for POWERON_RESET in the Ideas sub-forum ... I had some fun doing that on an ESP32.
davef
